src/test/unit/app/controller/courts/PostcodeController.ts

```diff
@@ -0,0 +1,622 @@
+    const req = mockRequest();
+
+    req.body = {
+      'existingPostcodes': getPostcodeInput,
+      'newPostcodes': '',
+      'csrfToken': CSRF.create(),
+      'courtTypes': courtTypesBodyInput,
+      'areasOfLaw': areasOfLawBodyInput
+    };
+    req.params = { slug: testSlug };
+    req.scope.cradle.api = mockApi;
+
+    await controller.post(req, res);
+
+    expect(res.render).toBeCalledWith('courts/tabs/postcodesContent', {
+      postcodes: getPostcodeData,
+      courts: [],
+      slug: testSlug,
+      searchValue: '',
+      updated: true,
+      errors: [{text: controller.noPostcodeErrorMsg}],
+      isEnabled: true,
+      areasOfLaw: areasOfLawMethodOutput,
+      courtTypes: courtTypesMethodOutput
+    });
+    expect(mockApi.addPostcodes).not.toBeCalled();
+    expect(mockApi.getCourtAreasOfLaw).not.toBeCalled();
+    expect(mockApi.getCourtCourtTypes).not.toBeCalled();
+  });
+
+  test('Should not delete postcodes if CSRF token is invalid', async() => {
+    const res = mockResponse();
+    const req = mockRequest();
+
+    CSRF.verify = jest.fn().mockReturnValue(false);
+    req.body = {
+      'existingPostcodes': getPostcodeInput,
+      'selectedPostcodes': newPostcodes,
+      'csrfToken': CSRF.create(),
+      'courtTypes': courtTypesBodyInput,
+      'areasOfLaw': areasOfLawBodyInput
+    };
+    req.params = { slug: testSlug };
+    req.scope.cradle.api = mockApi;
+
+    await controller.delete(req, res);
+
+    expect(res.render).toBeCalledWith('courts/tabs/postcodesContent', {
+      postcodes: getPostcodeData,
+      courts: [],
+      slug: testSlug,
+      searchValue: '',
+      updated: false,
+      errors: [{text: controller.addErrorMsg}],
+      isEnabled: true,
+      areasOfLaw: areasOfLawMethodOutput,
+      courtTypes: courtTypesMethodOutput
+    });
+    expect(mockApi.deletePostcodes).not.toBeCalled();
+    expect(mockApi.getCourtAreasOfLaw).not.toBeCalled();
+    expect(mockApi.getCourtCourtTypes).not.toBeCalled();
+  });
+
+  test('Should handle delete postcode no selection error', async() => {
+    const res = mockResponse();
+    const req = mockRequest();
+
+    req.body = {
+      'existingPostcodes': getPostcodeInput,
+      'selectedPostcodes': '',
+      'csrfToken': CSRF.create(),
+      'courtTypes': courtTypesBodyInput,
+      'areasOfLaw': areasOfLawBodyInput
+    };
+    req.params = { slug: testSlug };
+    req.scope.cradle.api = mockApi;
+
+    await controller.delete(req, res);
+
+    expect(res.render).toBeCalledWith('courts/tabs/postcodesContent', {
+      postcodes: getPostcodeData,
+      courts: [],
+      slug: testSlug,
+      searchValue: '',
+      updated: false,
+      errors: [{text: controller.noSelectedPostcodeMsg}],
+      isEnabled: true,
+      areasOfLaw: areasOfLawMethodOutput,
+      courtTypes: courtTypesMethodOutput
+    });
+    expect(mockApi.deletePostcodes).not.toBeCalled();
+    expect(mockApi.getCourtAreasOfLaw).not.toBeCalled();
```

Comment:
These 2 expectations for getCourtAreasOfLaw and getCourtCourtTypes are used in quite a few tests - are they needed?

Reply:
See other answer to this